benchmark/utils/generator.hpp

```diff
@@ -0,0 +1,278 @@
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GINKGO_BENCHMARK_UTILS_GENERATOR_HPP
+#define GINKGO_BENCHMARK_UTILS_GENERATOR_HPP
+
+
+#include <string>
+
+
+#include "benchmark/utils/formats.hpp"
+#include "benchmark/utils/general.hpp"
+#include "benchmark/utils/loggers.hpp"
+#include "benchmark/utils/stencil_matrix.hpp"
+#if GINKGO_BUILD_MPI
+#include "benchmark/utils/distributed_helpers.hpp"
+#endif
+
+
+template <typename ValueType = etype, typename IndexType = itype>
+struct DefaultSystemGenerator {
+    using index_type = IndexType;
+    using value_type = ValueType;
+    using Vec = vec<ValueType>;
+
+    static gko::matrix_data<ValueType, IndexType> generate_matrix_data(
+        rapidjson::Value& config)
+    {
+        if (config.HasMember("filename")) {
+            std::ifstream in(config["filename"].GetString());
+            return gko::read_generic_raw<ValueType, IndexType>(in);
+        } else if (config.HasMember("stencil")) {
+            return generate_stencil<ValueType, IndexType>(
+                config["stencil"].GetString(), config["size"].GetInt64());
+        } else {
+            throw std::runtime_error(
+                "No known way to generate matrix data found.");
+        }
+    }
+
+    static std::shared_ptr<gko::LinOp> generate_matrix(
+        std::shared_ptr<const gko::Executor> exec, rapidjson::Value& config)
+    {
+        auto data = generate_matrix_data(config);
+        return generate_matrix_with_format(
+            std::move(exec), config["optimal"]["spmv"].GetString(), data);
+    }
+
+    static std::shared_ptr<gko::LinOp> generate_matrix_with_format(
+        std::shared_ptr<const gko::Executor> exec,
+        const std::string& format_name,
+        const gko::matrix_data<ValueType, itype>& data)
+    {
+        return gko::share(
+            ::formats::matrix_factory(format_name, std::move(exec), data));
+    }
+
+    static std::shared_ptr<gko::LinOp> generate_matrix_with_default_format(
+        std::shared_ptr<const gko::Executor> exec,
+        const gko::matrix_data<ValueType, itype>& data)
+    {
+        return generate_matrix_with_format(std::move(exec), "coo", data);
+    }
+
+    static std::unique_ptr<Vec> create_multi_vector(
+        std::shared_ptr<const gko::Executor> exec, gko::dim<2> size,
+        ValueType value)
+    {
+        auto res = Vec::create(exec);
+        res->read(gko::matrix_data<ValueType, itype>(size, value));
+        return res;
+    }
+
+    static std::unique_ptr<Vec> create_multi_vector(
+        std::shared_ptr<const gko::Executor> exec, gko::dim<2> size,
+        gko::size_type stride)
+    {
+        auto res = Vec::create(exec, size, stride);
+        return res;
+    }
```

Comment:
These two overloads are a bit too close for comfort. If you forget a `.0` somewhere, you might inadvertently create an uninitialized vector. I would suggest a specific name
```suggestion
    static std::unique_ptr<Vec> create_multi_vector(
        std::shared_ptr<const gko::Executor> exec, gko::dim<2> size,
        ValueType value)
    {
        auto res = Vec::create(exec);
        res->read(gko::matrix_data<ValueType, itype>(size, value));
        return res;
    }

    static std::unique_ptr<Vec> create_multi_vector_strided(
        std::shared_ptr<const gko::Executor> exec, gko::dim<2> size,
        gko::size_type stride)
    {
        auto res = Vec::create(exec, size, stride);
        return res;
    }
```

Reply:
it's used in create_vector, see the thread below

Reply:
No wait, I'm not entirely sure which overload fits better, ValueType or size_type? How does this even compile? :laughing:

Reply:
Yes, that one below is definitely wrong :sweat_smile:. But I've found the actual use for that in the blas benchmark, so I will fix it.